File: exam/src/objects/Proposal.ts
import { Document } from "./Document";

export class Proposal extends Document {
    private proposalDate: string = "";
    private client: string = "";

    constructor (filename: string, extension: string, size: number, proposalDate: string, client: string) {
        super(filename, extension, size);
        this.proposalDate = proposalDate;
        this.client = client;
    }

    public getProposalDate(): string {
        return this.proposalDate;
    }

    public getClient(): string {
        return this.client;
    }

    public setProposalDate(proposalDate: string): void {
        this.proposalDate = proposalDate;
    }

    public setClient(client: string): void {
        this.client = client;
    }
}

File: exam/src/documents/Proposal.java
package documents;

import java.util.HashMap;

public class Proposal extends Document {

    public Proposal(String filename, HashMap<String, String> metadata) {
        super(filename, metadata);
    }

    @Override
    public boolean acceptedSize() {
        return Float.parseFloat(this.metadata.get("size")) < 5.00;
    }

    @Override
    public boolean acceptedExtensions() {
        return this.filename.toLowerCase().endsWith(".pdf") ||  this.filename.toLowerCase().endsWith(".docx");
    }

    @Override
    public boolean hasRequiredMetadata() {
        return this.metadata.containsKey("proposalDate") && this.metadata.containsKey("client");
    }

    @Override
    public boolean isValid() {
        return super.isValid() && acceptedSize() && acceptedExtensions() && hasRequiredMetadata();
    }
}
